src/circllhist.c

```diff
@@ -1637,3 +1637,49 @@ hist_approx_inverse_quantile(const histogram_t *hist, const double *in, int in_s
   }
   return 0;
 }
+
+histogram_t *
+hist_create_approximation_from_adhoc(histogram_approx_mode_t mode,
+                                     const histogram_adhoc_bin_t *bins,
+                                     size_t nbins, double sum) {
+  histogram_t *h = hist_alloc_nbins(nbins);
+  double last_spread = 0;
+  for(size_t i=0; i<nbins; i++) {
+    double a = bins[i].lower;
+    double b = bins[i].upper;
+    if(abs(a) > abs(b)) {
+      a = bins[i].upper;
+      b = bins[i].lower;
+    }
+    // chose a midpoint with minimal error
+    double m = 0;
+    switch(mode) {
+      case HIST_APPROX_HIGH:
+        m = bins[i].upper - fabs(bins[i].upper) / 1000;
+        break;
+      case HIST_APPROX_LOW:
+        m = bins[i].lower + fabs(bins[i].lower) / 1000;
+        break;
+      case HIST_APPROX_MID:
+        m = a + (b-a)/2;
+        break;
+      case HIST_APPROX_MIN_ERROR:
+        if(a == 0 || b == 0)
+          m = a + (b-a)/2;
+        else
+          m = (a / (a + b)) * fabs(b-a) + bins[i].lower;
+        break;
+    }
+    // however if we're a (-inf,B) or an (A,+inf) bucket we need to be more clever
+    if(i>0 && bins[i].upper >= pow(10,128)) {
+      m = a + (last_spread > a ? a : last_spread);
+    }
+    if(i == 0 && nbins > 2 && bins[0].lower <= -pow(10,128)) {
+      double offset1 = abs(b), offset2 = abs(bins[1].upper - bins[1].lower);
+      m = b - (offset1 > offset2 ? offset2 : offset1);
+    }
```

Comment:
Not sure about those special cases. (I don't see a problem, but I it's not clear to me what this is doing).

Reply:
This is the handle the case where one end of the interval is 0 or 0-like.  The harmonic mean in that case is not benefiting the user... 0.